src/lightgbm/src/main/scala/TrainUtils.scala

```diff
@@ -36,95 +36,172 @@ private object TrainUtils extends Serializable {
+          result + " and is finished: " + isFinished)
+      } catch {
+        case e: java.lang.Exception => {
+          isFinished = 1
+          log.info("LightGBM reached early termination on one of the workers, stopping training on other workers")
+        }
+      }
+      if (hasValid && isFinished == 0) {
+        val evalResults = lightgbmlib.new_doubleArray(bestEval.get.evalCounts)
+        val dummyEvalCountsPtr = lightgbmlib.new_intp()
+        val resultEval = lightgbmlib.LGBM_BoosterGetEval(boosterPtr.get, 1, dummyEvalCountsPtr, evalResults)
+        lightgbmlib.delete_intp(dummyEvalCountsPtr)
+        LightGBMUtils.validate(resultEval, "Booster Get Eval")
+        (0 until bestEval.get.evalCounts).foreach { index =>
+          val score = lightgbmlib.doubleArray_getitem(evalResults, index)
+          val evalName = bestEval.get.evalNames(index)
+          val cmp =
+            if (evalName.startsWith("auc") || evalName.startsWith("ndcg@") || evalName.startsWith("map@"))
+              (x: Double, y: Double) => x > y
+            else (x: Double, y: Double) => x < y
+          if (bestEval.get.bestScores(index).isEmpty || cmp(score, bestEval.get.bestScore(index))) {
+            bestEval.get.bestScore(index) = score
+            bestEval.get.bestIter(index) = iters
+            bestEval.get.bestScores(index) = (0 until bestEval.get.evalCounts)
+              .map(j => lightgbmlib.doubleArray_getitem(evalResults, j)).toArray
+          } else if (iters - bestEval.get.bestIter(index) >= trainParams.earlyStoppingRound) {
+            isFinished = 1
+            log.info("Early stopping, best iteration is " + bestEval.get.bestIter(index))
+          }
         }
+        lightgbmlib.delete_doubleArray(evalResults)
+      }
+      iters = iters + 1
+    }
+  }
 
-        val isFinishedPtr = lightgbmlib.new_intp()
-        var isFinised = 0
-        var iters = 0
-        while (isFinised == 0 && iters < trainParams.numIterations) {
-          val result = lightgbmlib.LGBM_BoosterUpdateOneIter(boosterPtr.get, isFinishedPtr)
-          LightGBMUtils.validate(result, "Booster Update One Iter")
-          isFinised = lightgbmlib.intp_value(isFinishedPtr)
-          log.info("LightGBM running iteration: " + iters + " with result: " +
-            result + " and is finished: " + isFinised)
-          iters = iters + 1
+  def translate(labelColumn: String, featuresColumn: String, weightColumn: Option[String],
+                validationData: Option[Array[Row]], log: Logger,
+                trainParams: TrainParams, inputRows: Iterator[Row]): Iterator[LightGBMBooster] = {
+    if (!inputRows.hasNext)
+      List[LightGBMBooster]().toIterator
+    val rows = inputRows.toArray
+    var trainDatasetPtr: Option[SWIGTYPE_p_void] = None
+    var validDatasetPtr: Option[SWIGTYPE_p_void] = None
+    try {
+      val (trainRows, validRows) =
+        if (validationData.isDefined) {
```

Comment:
val (trainRows, validRows) = (rows, validationData)

isnt this what you are doing?

Reply:
good catch, I think I refactored this several times and forgot to remove this